src/rpc/blockchain.cpp

```diff
@@ -1531,6 +1538,313 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+    }
+}
+
+template<typename T>
+static T CalculateTruncatedMedian(std::vector<T>& scores)
+{
+    size_t size = scores.size();
+    if (size == 0) {
+        return 0;
+    } if (size == 1) {
+        return scores[0];
+    }
+
+    std::sort(scores.begin(), scores.end());
+    if (size % 2 == 0) {
+        return (scores[size / 2 - 1] + scores[size / 2]) / 2;
+    } else {
+        return scores[size / 2];
+    }
+}
+
+// outpoint (needed for the utxo index) + nHeight + fCoinBase
+static const size_t PER_UTXO_OVERHEAD = sizeof(COutPoint) + sizeof(uint32_t) + sizeof(bool);
+
+static void UpdateBlockStats(const CBlockIndex* pindex, std::set<std::string>& stats, std::map<std::string, UniValue>& map_stats)
+{
+    int64_t inputs = 0;
+    int64_t outputs = 0;
+    int64_t swtxs = 0;
+    int64_t total_size = 0;
+    int64_t total_weight = 0;
+    int64_t swtotal_size = 0;
+    int64_t swtotal_weight = 0;
+    int64_t utxo_size_inc = 0;
+    CAmount total_out = 0;
+    CAmount totalfee = 0;
+    CAmount minfee = MAX_MONEY;
+    CAmount maxfee = 0;
+    CAmount minfeerate = MAX_MONEY;
+    CAmount maxfeerate = 0;
+    CAmount minfeerate_old = MAX_MONEY;
+    CAmount maxfeerate_old = 0;
+    std::vector<CAmount> fee_array;
+    std::vector<CAmount> feerate_old_array;
+    std::vector<CAmount> feerate_array;
+
+    CBlock block;
+    ReadBlockCheckPruned(block, pindex);
+
+    for (const auto& tx : block.vtx) {
+        outputs += tx->vout.size();
+        CAmount tx_total_out = 0;
+        for (const CTxOut& out : tx->vout) {
+            utxo_size_inc += GetSerializeSize(out, SER_NETWORK, PROTOCOL_VERSION) + PER_UTXO_OVERHEAD;
+            tx_total_out += out.nValue;
+        }
+
+        if (tx->IsCoinBase()) {
+            continue;
+        }
+        total_out += tx_total_out;
+        inputs += tx->vin.size(); // Don't count coinbase's fake input
+        int64_t tx_size = tx->GetTotalSize();
+        total_size += tx_size;
+        int64_t weight = GetTransactionWeight(*tx);
+        total_weight += weight;
+
+        if (tx->HasWitness()) {
+            ++swtxs;
+            swtotal_size += tx_size;
+            swtotal_weight += weight;
+        }
+
+        CAmount tx_total_in = 0;
+        for (const CTxIn& in : tx->vin) {
+            CTransactionRef tx_in;
+            RpcGetTx(in.prevout.hash, tx_in);
+            CTxOut prevoutput = tx_in->vout[in.prevout.n];
+
+            tx_total_in += prevoutput.nValue;
+            utxo_size_inc -= GetSerializeSize(prevoutput, SER_NETWORK, PROTOCOL_VERSION) + PER_UTXO_OVERHEAD;
+        }
+        CAmount txfee = tx_total_in - tx_total_out;
+        assert(MoneyRange(txfee));
+        fee_array.push_back(txfee);
+        totalfee += txfee;
+        minfee = std::min(minfee, txfee);
+        maxfee = std::max(maxfee, txfee);
+
+        CAmount feerate_old = CFeeRate(txfee, tx_size).GetTruncatedFee(1);
```

Comment:
For feerate you should use the virtual size, not total size.

Reply:
feerate is using vsize, feerate_old is using old size. Perhaps we just want to remove the whole *_old family.